libnd4j/include/ops/declarable/helpers/cpu/lup.cpp

```diff
@@ -155,14 +172,112 @@ namespace helpers {
         if (swapCount % 2) determinant = -determinant;
         if (compound != nullptr)
             compound->assign(compoundMatrix);
-        if (permutation != nullptr)
-            permutation->assign(permutationMatrix);
+        if (permutation != nullptr) {
+            auto permutaionVector = NDArrayFactory::create('c', {rowNum}, DataTypeUtils::fromT<I>(), input->getContext());
+            for (auto i = 0; i < rowNum; i++) {
+                for (auto j = 0; j < columnNum; j++) {
+                    if (permutationMatrix.t<T>(i, j) != 0) {
+                        permutaionVector.template t<I>(i) = j;
+                    }
+                }
+            }
+            if (permutationMatrix.isSameShape(permutation))
+                permutation->assign(permutationMatrix);
+            else if (permutation->isSameShape(permutaionVector)) {
+                permutation->assign(permutaionVector);
+            }
+        }
         return determinant;
     }
 
-    BUILD_SINGLE_TEMPLATE(template NDArray lup_, (LaunchContext *context, NDArray* input, NDArray* output, NDArray* permutation), FLOAT_TYPES);
+    BUILD_DOUBLE_TEMPLATE(template NDArray lup_, (LaunchContext *context, NDArray* input, NDArray* output, NDArray* permutation), FLOAT_TYPES, INDEXING_TYPES);
+    /*
+     * lu decomposition with naive algorithm with partial pivoting
+     * */
+    template <typename T, typename I>
+    static I argmaxCol(I column, T* compoundBuffer, Nd4jLong* compoundShape) {
+        auto rowNum = shape::sizeAt(compoundShape, 0);
+        Nd4jLong xInitial[] = {column, column};
+        auto xInitialIndex = shape::getOffset(compoundShape, xInitial, 0);
+        auto maxValue = T(0); //nd4j::math::nd4j_abs(compoundBuffer[xInitialIndex]);
+        auto result = -1;
+        auto loop = PRAGMA_THREADS_FOR {
+            for (auto rowCounter = start; rowCounter < stop; rowCounter += increment) {
+                Nd4jLong xPos[] = {rowCounter, column};
+                auto xIndex = shape::getOffset(compoundShape, xPos, 0);
+                if (nd4j::math::nd4j_abs(compoundBuffer[xIndex]) > maxValue) {
+                    maxValue = nd4j::math::nd4j_max(maxValue, nd4j::math::nd4j_abs(compoundBuffer[xIndex]));
```

Comment:
maxValue is defined out of the function, so it will be shared by multiple threads. So you'll definitely have race condition here. Please fix this.